lms/security.py

```diff
@@ -119,10 +191,13 @@ def _permits(policy, request, _context, permission):
     if identity and permission in identity.permissions:
         return Allowed("allowed")
 
+    if identity and identity.denied_reason:
```

Comment:
As far as I can tell only permits can return one of these so we have to include it on identity and then expose it here.

Reply:
I wonder if maybe the security policy needs to save the `denied_reason` on its own `self`? That would avoid the need to put a `denied_reason` on the `Identity` class and would allow `identity` to be `None` when no user can be identified.

But we may just need to refactor our code a bit first to enable it.

If we split it up so that there's one security policy for each way of authenticating the user (LTI 1.1 launch params, LTI 1.3 params, bearer token, OAuth state param, Google, ...) and so that every request is always mapped to exactly one security policy according to the request path.

```python
class LTI1SecurityPolicy:
    def identity(self):
        try:
            return self._get_identity()
        except ValidationError as err:
            self.reason = err
            return None

    def permits(self):
        if identity is None:
            return DeniedWithReason(self.reason)
        ...

    def _get_identity(self):
        # Return an Identity from LTI 1.1 launch params or raise ValidationError.
        ...

    ...
```